File: src/main/java/org/jukeboxmc/block/BlockCocoa.java
package org.jukeboxmc.block;

import org.jukeboxmc.block.direction.Direction;
import org.jukeboxmc.item.ItemCocoa;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockCocoa extends Block {

    public BlockCocoa() {
        super( "minecraft:cocoa" );
    }

    @Override
    public ItemCocoa toItem() {
        return new ItemCocoa();
    }

    @Override
    public BlockType getBlockType() {
        return BlockType.COCOA;
    }

    @Override
    public boolean isSolid() {
        return false;
    }

    public void setAge( int value ) { // 0-15
        this.setState( "age", value );
    }

    public int getAge() {
        return this.stateExists( "age" ) ? this.getIntState( "age" ) : 0;
    }

    public void setDirection( Direction direction ) {
        switch ( direction ) {
            case SOUTH:
                this.setState( "direction", 0 );
                break;
            case WEST:
                this.setState( "direction", 1 );
                break;
            case NORTH:
                this.setState( "direction", 2 );
                break;
            case EAST:
                this.setState( "direction", 3 );
                break;
        }
    }

    public Direction getDirection() {
        int value = this.stateExists( "direction" ) ? this.getIntState( "direction" ) : 0;
        switch ( value ) {
            case 0:
                return Direction.SOUTH;
            case 1:
                return Direction.WEST;
            case 2:
                return Direction.NORTH;
            default:
                return Direction.EAST;
        }
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockCocoa.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.Cocoa
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.block.data.Direction
import org.jukeboxmc.api.item.Item
import org.jukeboxmc.api.item.ItemType
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

class BlockCocoa(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    Cocoa {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        this.setDirection(player.getDirection().opposite())
        return super.placeBlock(player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace)
    }

    override fun getDirection(): Direction {
        return when (this.getIntState("direction")) {
            0 -> Direction.EAST
            1 -> Direction.SOUTH
            2 -> Direction.WEST
            else -> Direction.NORTH
        }
    }

    override fun setDirection(value: Direction): BlockCocoa {
        return when (value) {
            Direction.EAST -> this.setState("direction", 0)
            Direction.SOUTH -> this.setState("direction", 1)
            Direction.WEST -> this.setState("direction", 2)
            else -> this.setState("direction", 3)
        }
    }

    override fun getAge(): Int {
        return this.getIntState("age")
    }

    override fun setAge(value: Int): BlockCocoa {
        return this.setState("age", value)
    }

    override fun getDrops(item: Item): MutableList<Item> {
        return if (this.getAge() != 3) {
            mutableListOf()
        } else {
            mutableListOf(Item.create(ItemType.COCOA_BEANS, 3))
        }
    }
}